algorithms/src/Kokkos_Algo.hpp

```diff
@@ -0,0 +1,111 @@
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_ALGO_HPP
+#define KOKKOS_ALGO_HPP
+
+#include <Kokkos_Core.hpp>
+
+/// \file Kokkos_Algo.hpp
+/// \brief Kokkos counterparts for Standard C++ Library algorithms
+
+namespace Kokkos {
+namespace Experimental {
+
+template <class DataType, class... Properties>
+auto begin(Kokkos::View<DataType, Properties...>& v) -> decltype(v.data()) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return v.data();
+}
+
+template <class DataType, class... Properties>
+auto begin(const Kokkos::View<DataType, Properties...>& v)
+    -> decltype(v.data()) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return v.data();
+}
+
+template <class DataType, class... Properties>
+auto end(Kokkos::View<DataType, Properties...>& v) -> decltype(v.data()) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return v.data() + v.size();
+}
+
+template <class DataType, class... Properties>
+auto end(const Kokkos::View<DataType, Properties...>& v) -> decltype(v.data()) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return v.data() + v.size();
+}
+
+template <class PointerType, class FunctorType>
+FunctorType for_each(PointerType data, PointerType end, FunctorType functor) {
+  const auto numOfElements = end - data;
+  Kokkos::parallel_for(
+      numOfElements, KOKKOS_LAMBDA(const int i) {
```

Comment:
you imply that this would help for bebugging purposes?

Reply:
Sounds useful, but if we do that we should probably add such overloads for all std algorithms that end up calling parallel operators.

Reply:
yeah certainly, if we do it we need to add it to all of them. Note that if we use functors for each of these guys we would by default get half way reasonable names since it uses typeids. The typeid of lambdas are shitty however. 

```
template<class PointerType, FunctorType>
struct ForEach {
  FunctorType m_f;
  ...
}

template<class PointerType, FunctorType>
FunctorType for_each(PointerTYpe data, ...) {
   parallel_for(N,ForEach(data,functor));
}
```
would result in a name like Kokkos;:Impl::ForEach<PointerType,FunctorType> after demangling.
So not too terrible.